docs/source/reference/fuzzer_configuration.md
# Fuzzer Configuration

The kAFL fuzzer configuration can be changed via config files, environment variables and command line switches.

The [Dynaconf](https://www.dynaconf.com/) framework is used behind the scenes for configuration management, so
everything that you learn from Dynaconf's documentation should also be applicable for kAFL fuzzer.

## Configuration sources and precedence

All configuration files are using the `YAML` format.

1. kAFL fuzzer packaged default configuration file: [`kafl_fuzzer/common/config/default_settings.yaml`](https://github.com/IntelLabs/kafl.fuzzer/blob/master/kafl_fuzzer/common/config/default_settings.yaml)
2. System-level config file: `/etc/xdg/kAFL/settings.yaml`
3. User-level config file: `$HOME/.config/kAFL/settings.yaml`
4. Local config file: `$PWD/kafl.yaml`
5. Environment variable `KAFL_CONFIG_FILE` if specified
6. Command line arguments. Ex: `--work_dir /dev/shm/kafl_test_feature`
7. Configuration keys specified by environment variables prefixed by `KAFL_`. Ex: `KAFL_PROCESSES=8`

Note: if `KAFL_CONFIG_FILE` is specified but points to a non-existing file, a validation error will be raised (to warn the user).

## Overriding settings from environment variables

You can override any setting key by exporting an environment variable prefixed by `KAFL_`.

Example

~~~shell
export KAFL_QEMU_MEMORY=1024
export KAFL_LOG_HRPINTF=TRUE
~~~


## Configuration keys

The following section lists all configuration keys avaialble in kAFL, and their corresponding command line switches, if any.

:::{note}
Configuration keys are case insensitive:
~~~YAML
qemu_memory: 256
# and
QEMU_MEMORY: 256
~~~
:::

### `abort_exec`

Exit kAFL fuzzing a after `<n>` total executions.

:::{note}
Default: `0`

Command-line: `--abort-exec`
:::

### `abort_time`

Builtin timeout to stop kafl fuzzing after `<n>` seconds elapsed.

:::{note}
Default: `0`

Command-line: `--abort-time`
:::

### `action`

kAFL debug action.

Choices available:

| Action          | Description                                              |
| --------------- | -------------------------------------------------------- |
| `benchmark`     | Perform performance benchmark                            |
| `gdb`           | Run payload with GDB server (must compile with redqueen) |
| `trace`         | Peform trace run                                         |
| `trace-qemu`    | Perform trace run and print QEMU stdout                  |
| `noise`         | Perform run and measure non determinism                  |
| `printk`        | redirect printk calls to kAFL                            |
| `redqueen`      | Run redqueen debugger                                    |
| `redqueen-qemu` | Run redqueen debugger and print QEMU stdout              |
| `verify`        | Run verification steps                                   |

:::{note}
Command-line: `--action`.
:::

### `afl_arith_max`

Maximum number of increment/decrement steps in AFL-style arithmetic mutation
(only affects deterministic stage, not havoc).

:::{note}
Default: 34

Command-line: `--afl-arith-max`
:::

### `afl_dumb_mode`

Skip AFL-style deterministic stages (`bitflip`, `arithmetic`, `interesting` mutations).

:::{note}
Default: `false`

Command-line: `--afl-dumb-mode`, `-D`
:::

### `afl_skip_zero`

Skip mutating zero-bytes in AFL-style deterministic stages.

:::{note}
Default: `false`

Command-line: `--afl-skip-zero`
:::

### `bitmap_size`

Size of feedback bitmap (must be power of 2).

:::{note}
Default: `65536`

Command-line: `--bitmap-size`
:::

### `cpu_offset`

Modify automated CPU pinning to start at vCPU `n` and assign Qemu/Worker
instances to the next [`-p`](#processes) vCPUs.

:::{note}
Default: `0`

Command-line: `--cpu-offset <n>`
:::

### `debug`

Enable additional (expensive) debug messages during execution.

- switches logging to `DEBUG` level
- toggles QEMU `nyx` log item (`-d nyx`), only effective if [`--log`](#log) has been specified
- stores funky input data in `$WORKDIR/funky`. See [`--funky`](#funky) and [workdir reference](./workdir_layout.md)

Implies [`--verbose`](#verbose).

:::{note}
Default: `false`

Command-line: `--debug`
:::

### `dict`

Use file `<path>` as source of dictionary inputs in havoc stage.

See the examples at [`kafl.fuzzer/dict`](https://github.com/IntelLabs/kafl.fuzzer/tree/master/dict) for file format.

:::{note}
Default: `None`

Command-line: `--dict <path>`
:::

### `funky`

::::{tab-set}
:::{tab-item} kafl fuzz

Validate new (non-crashing/non-timeout) inputs multiple times and add to corpus
if consistent new coverage is found on majority of re-runs.  Executes payloads
8 times by default and accepts payloads that reproduce with 75% probability. In
combination with `--debug`, non-reproducible inputs are stored in
`$KAFL_WORKDIR/funky/`. Useful with non-deterministic targets.

:::
:::{tab-item} kafl cov
Output any additional found coverage and payloads in combination with `--trace --action noise`.
:::
::::

:::{note}
Default: `false`

Command-line: `--funky`
:::

### `gdbserver`

Starts QEMU GDB server (Extends QEMU command line with `-s -S`).

From QEMU command line help:
- `-s`: shorthand for -gdb tcp::1234
- `-S`: freeze CPU at startup (use 'c' to start execution)

:::{note}
Default: `false`

Command-line: `--gdbserver`
:::

### `grimoire`

Enable the [Grimoire](https://github.com/RUB-SysSec/grimoire) fuzzer stage.

:::{note}
Default: `false`

Command-line: `--grimoire`
:::


### `input`

::::{tab-set}
:::{tab-item} kafl cov
Sets the kAFL workdir to be processed for coverage information. Will
automatically process files in `corpus/{crash,kasan,regular}/` (but not `timeout`).
:::
:::{tab-item} kafl debug
Sets the payload file to be used as input for the debugging session.
:::
::::

:::{note}
Default: [workdir](#work_dir) value

Command-line: `--input <path>`
:::

### `ip0-1-2-3`

Set the IP filter ranges (code ranges) to be used with Intel PT. The filter
ranges may also be set by the agent using the (`RANGE_SUBMIT` hypercall)[hypercall_api.md].

Not setting a IP filter region is currently not supported.

:::{note}
Default: `None`

Command-line: `--ip0`, `--ip1`, `--ip2`, `--ip3`
:::

### `iterations`

Execute the debugged payload `<n>` times.

Used by `noise`, `trace` and `trace-qemu` debug [actions](#action) in case of
non-deterministic execution.

:::{note}
Default: `5`

Command-line: `--iterations`

Applicable subcommands: `debug`
:::

### `kickstart`

When no payloads in queue or more workers than available corpus payloads,
let the worker kickstart fuzzing with random strings of length `n`.
Useful to test a new harness that has no well defined seeds.

Enabled by default. Set `0` to disable.

:::{note}
Default: `256`

Command-line: `--kickstart <n>`
:::

### `log`

Add an additional file logging handler to `$KAFL_WORKDIR/kafl_fuzzer.log`.
Also redirects Qemu log output to `$KAFL_WORKDIR/qemu_trace_NN.log`, if any.

:::{note}
Default: `false`

Command-line: `--log`
:::

### `log_crashes`

Like `log_hprintf`, but copy the hprintf log to `$KAFL_WORKDIR/logs/` for any
new found `{crash,kasan,timeout}` type payload.  Also truncates the main hprintf
log after every execution.

This is the recommended guest logging option to collect live guest logs
corresponding to new found non-regular payloads, while avoiding OOM due to huge
hprintf logs.

:::{note}
Default: `false`

Command-line: `--log-crashes`
:::

### `log_hprintf`

Redirect guest `hprintf` output to `$KAFL_WORKDIR/hprintf_NN.log` (see [PRINTF hypercall](hypercall_api.md)).

Creates a linear log of Guest execution across snapshot/restore. Recommended for
debugging guest/harness execution.

:::{note}
Default: `false`

Command-line: `--log-hprintf`
:::

### `payload_size`

Maximum payload size in bytes. Must be multiple of page size (4096 bytes).

:::{note}
Default: `131072`

Command-line: `--payload-size <n>`
:::

### `processes`

Number of processes to launch for parallelized fuzzing and coverage.

:::{note}
Default: `1`

Command-line: `--processes <n>`
:::

### `ptdump_path`

Path to ptdump executable.

:::{note}
Default: `$LIBXDC_ROOT/build/ptdump_static`

Command-line: `--ptdump-path`
:::

### `purge`

Purge the [KAFL workdir](#work_dir) directory at startup.

:::{note}
Default: `false`

Command-line: `--purge`
:::

### `qemu_append`

Kernel command line if [`--kernel`](#qemu_kernel) is used.

Corresponds to QEMU `-append <cmdline>` argument.

:::{note}
Default: `nokaslr oops=panic nopti mitigations=off console=ttyS0`

Command-line: `--qemu-append`
:::

### `qemu_base`

Baseline for QEMU command-line.

:::{note}
Default: `-enable-kvm -machine kAFL64-v1 -cpu kAFL64-Hypervisor-v1,+vmx -no-reboot -net none -display none`

Command-line: `--qemu-base`
:::

### `qemu_bios`

Set custom BIOS image to be used by QEMU.

Corresponds to QEMU `-bios <file>` argument.

:::{note}
Default: `None`

Command-line: `--bios`
:::

### `qemu_extra`

Extra string appended to QEMU command line. Useful to override specific Qemu flags.

:::{note}
Default: `None`

Command-line: `--qemu-extra`
:::

### `qemu_image`

Path to Qemu disk image.

Corresponds to QEMU `-drive file=<disk>` argument.

:::{note}
Default: `None`

Command-line: `--image`
:::

### `qemu_initrd`

Initial ram disk for QEMU.

Corresponds to QEMU `-initrd <file>` argument.

:::{note}
Default: `None`

Command-line: `--initrd`
:::

### `qemu_kernel`

bzImage to use as kernel image for QEMU.

Corresponds to QEMU `-kernel <image>` argument.

:::{note}
Default: `None`

Command-line: `--kernel`
:::

### `qemu_memory`

Amount of memory in `MB` for QEMU.

Corresponds to QEMU `-m <size>` argument.

:::{note}
Default: `256`

Command-line: `--memory <n>`, `-m <n>`
:::

### `qemu_path`

Path to QEMU executable with Nyx patches.

:::{note}
Default: `$QEMU_ROOT/x86_64-softmmu/qemu-system-x86_64`

Command-line: `--qemu-path <path>`
:::

### `qemu_serial`

Extend QEMU command line with the configuration value, and then append
`-chardev file,id=kafl_serial,mux=on,path=$KAFL_WORKDIR/serial_<qemu_pid>.log`.

:::{note}
Default: `-device isa-serial,chardev=kafl_serial`

Command-line: `--qemu-serial`
:::

### `qemu_snapshot`

Path to VM pre-snapshot directory. Use with manual `pre-snapshot` creation (`LOCK` hypercall)[hypercall_api.md]

:::{note}
Default: `None`

Command-line: `--snapshot <path>`
:::

### `quiet`

Set Python [`logging`](https://docs.python.org/3/library/logging.html) level to [`WARNING`](https://docs.python.org/3/library/logging.html#logging-levels).

:::{note}
Default: `false`

Command-line: `--quiet`
:::

### `radamsa_path`

Path to radamsa executable.

:::{note}
Default: `$RADAMSA_ROOT/bin/radamsa`

Command-line: `--radamsa-path <path>`
:::

### `radamsa`

Enable radamsa mutation as part of `havoc` stage.

:::{note}
Default: `false`

Command-line: `--radamsa`
:::

### `redqueen_simple`

Modify redqueen to also process 'simple' inputs that would likely be found by
subsequent deterministic/havoc mutations.

:::{note}
Default: `false`

Command-line: `--redqueen-simple`
:::

### `redqueen_hammer`

Enable Redqueen jump table hammering.

:::{note}
Default: `false`

Command-line: `--redqueen-hammer`
:::

### `redqueen_hashes`

Enable Redqueen checksum fixer (broken).

:::{note}
Default: `false`

Command-line: `--redqueen-hashes`
:::

### `redqueen`

Enable Redqueen analysis + mutation stages.

:::{note}
Default: `false`

Command-line: `--redqueen`
:::

### `reload`

Reload the snapshot only on non-regular executions or every `<N>` regular executions.
Set to `0` for fully persistent mode execution.
Set to '>1' for partial persistent mode execution.

For targets that support persistent mode, setting '--reload' somewhere in the
range of 10-100 tends to yield good performance/stability trade-off.

:::{note}
Default: `1`

Command-line: `--reload <n>`, `-R <n>`
:::

### `resume`

Tell kAFL to resume operation based on an existing workdir. This is currently
limited to `kafl cov` and `kafl debug`, where it will cause the Qemu instance to
initialize based on existing Nyx `snapshot/` and `page_cache.*` files.

Resuming a complete fuzzer campaign (kafl fuzz) is not yet supported.

:::{note}
Default: `false`

Command-line: `--resume`
:::

### `seed_dir`

Specify a directory `<path>` to use as seed directory. The directory is
traversed recursively and files are copied to the kAFL
[workdir](../reference/workdir_layout.md) as `$KAFL_WORKDIR/imports/seed_xxx`,
where there are consumed upon fuzzer startup.

See the examples at [`kafl.fuzzer/seed`](https://github.com/IntelLabs/kafl.fuzzer/tree/master/seed)

:::{note}
Default: `None`

Command-line: `--seed-dir <path>`
:::

### `sharedir`

Path to the Qemu 'sharedir' directory. Files in this folder will be made
available for the agent to download via (`REQ_STREAM_DATA` hypercall)[hypercall_api.md].

Appends `,sharedir=<value>` to the `nyx` QEMU device.

:::{note}
Default: `None`

Command-line: `--sharedir <path>`
:::

### `timeout_check`

When using both `timeout_hard` and `timeout_soft`, check that a payload reported as
'timeout' also produces a timeout when executing with maximum timeout 'timeout_hard'.

:::{note}
Default: `false`

Command-line: `--t-check`
:::

### `timeout_hard`

Hard execution timeout (seconds)

:::{note}
Default: `4`

Command-line: `--t-hard`
:::

### `timeout_soft`

Soft execution timeout (seconds). Used as initial lower timeout for Qemu
instances and then adapted based on seen performance, up to the maximum of
`timeout_hard`.

:::{note}
Default: `0.001`

Command-line: `--t-soft`
:::

### `trace_cb`

Run Qemu in 'callback-trace' mode, where PT traces are directly decoded using
a libxdc callback function. This tends to significantly slow down execution and
results in different coverage bitmaps. Deprecated.

Appends `,edge_cb_trace` to the `nyx` QEMU device.

:::{note}
Default: `false`

Command-line: `--trace-cb`
:::

### `trace`

Run Qemu in `dump-trace` mode, where binary PT traces are directly stored to
the workdir by each Qemu process. For any new discovered payloads, kAFL stores
the corresponding binary traces to `$KAFL_WORKDIR/traces/` for later decoding
with `kafl cov`.
This is the recommeded trace mode as it incurrs minimal slow-down and also
works for non-deterministic targets.

Appends `,dump_pt_trace` to the `nyx` QEMU device.

:::{note}
Default: `false`

Command-line: `--trace`
:::

### `verbose`

Enable verbose output by setting Python [`logging`](https://docs.python.org/3/library/logging.html) level to [`DEBUG`](https://docs.python.org/3/library/logging.html#logging-levels).

This option is also useful to dump and debug kafl configuration at load-time.

Example with [Fuzzing the Linux kernel](../tutorials/linux/fuzzing_linux_kernel.md) tutorial, if `--verbose` switch were to be added to the command line:

First kAFL will dump the list of loaded configuration files.
Check the [load order](#configuration-sources-and-precedence)
~~~
Loaded configuration files:
['/home/mtarral/kAFL/kafl/fuzzer/kafl_fuzzer/common/config/default_settings.yaml',
 '/home/mtarral/kAFL/kafl/examples/linux-kernel/./kafl_config.yaml']
~~~

Then the command-line configuration as parser by [argparse](https://docs.python.org/3/library/argparse.html)
~~~
Command line configuration:
{'abort_exec': None,
 'abort_time': None,
 'afl_arith_max': None,
 'afl_dumb_mode': True,
...
~~~

Followed by the configuration values extracted from the loaded config files:
~~~
Loaded configuration values:
{'env_global': {'CONFIG_FILE': './kafl_config.yaml',
                'ROOT': '/home/mtarral/kAFL/kafl',
                'WORKDIR': '/dev/shm/kafl_mtarral',
                'WORKSPACE': '/home/mtarral/kAFL'},
 'yaml': {'AFL_ARITH_MAX': 34,
          'BITMAP_SIZE': 65536,
          'CPU_OFFSET': 0,
          'DEBUG': False,
...
~~~

As the command line supersedes the configuration files, kAFL will override these values and finally dump the definitive configuration:
~~~
Final configuration:
{'ABORT_EXEC': 0,
 'ABORT_TIME': 0.0,
 'AFL_ARITH_MAX': 34,
 'AFL_DUMB_MODE': True,
~~~

:::{note}
Default: `false`

Command-line: `--verbose`
:::

### `work_dir`

The workdir is used by kAFL to accumulate results and communicate with QEMU and other processes.
It is the primary location for inspecting the status of a previous or still running kAFL session.

Also, any post-processing steps such as triage or coverage analysis typically builds up on an existing workdir.

:::{note}
Default: `/dev/shm/kafl_$USER`

Command-line: `--workdir`
:::

:::{seealso}
See also the [Workdir Layout](./workdir_layout.md) reference.
:::


kafl_fuzzer/common/config.py
# Copyright 2017-2019 Sergej Schumilo, Cornelius Aschermann, Tim Blazytko
# Copyright 2019-2020 Intel Corporation
#
# SPDX-License-Identifier: AGPL-3.0-or-later

import argparse
import os
import re
import sys

import confuse
from flatdict import FlatDict

from kafl_fuzzer.common.util import is_float, is_int, Singleton
from kafl_fuzzer.common.logger import logger




class FullPath(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        setattr(namespace, self.dest, os.path.abspath(os.path.expanduser(values)))


def create_dir(dirname):
    if not os.path.isdir(dirname):
        try:
            os.makedirs(dirname)
        except:
            msg = "Cannot create directory: {0}".format(dirname)
            raise argparse.ArgumentTypeError(msg)
    return dirname


def parse_is_dir(dirname):
    if not os.path.isdir(dirname):
        msg = "{0} is not a directory".format(dirname)
        raise argparse.ArgumentTypeError(msg)
    else:
        return dirname


def parse_is_file(dirname):
    if not os.path.isfile(dirname):
        msg = "{0} is not a file".format(dirname)
        raise argparse.ArgumentTypeError(msg)
    else:
        return dirname


def parse_ignore_range(string):
    m = re.match(r"(\d+)(?:-(\d+))?$", string)
    if not m:
        raise argparse.ArgumentTypeError("'" + string + "' is not a range of number.")
    start = min(int(m.group(1)), int(m.group(2)))
    end = max(int(m.group(1)), int(m.group(2))) or start
    if end > (128 << 10):
        raise argparse.ArgumentTypeError("Value out of range (max 128KB).")

    if start == 0 and end == (128 << 10):
        raise argparse.ArgumentTypeError("Invalid range specified.")
    return list([start, end])


def parse_range_ip_filter(string):
    m = re.match(r"([(0-9abcdef]{1,16})(?:-([0-9abcdef]{1,16}))?$", string.replace("0x", "").lower())
    if not m:
        raise argparse.ArgumentTypeError("'" + string + "' is not a range of number.")

    # print(m.group(1))
    # print(m.group(2))
    start = min(int(m.group(1).replace("0x", ""), 16), int(m.group(2).replace("0x", ""), 16))
    end = max(int(m.group(1).replace("0x", ""), 16), int(m.group(2).replace("0x", ""), 16)) or start

    if start > end:
        raise argparse.ArgumentTypeError("Invalid range specified.")
    return list([start, end])

def hidden(msg, unmask=False):
    if unmask or 'KAFL_CONFIG_DEBUG' in os.environ:
        return msg
    return argparse.SUPPRESS

# General startup options used by fuzzer, qemu, and/or utilities
def add_args_general(parser):
    parser.add_argument('-h', '--help', action='help',
                        help='show this help message and exit')
    parser.add_argument('-w', '--work-dir', metavar='<dir>', action=FullPath, type=str,
                        required=True, help='path to the output/working directory.')
    parser.add_argument('--purge', required=False, help='purge the working directory at startup.',
                        action='store_true', default=False)
    parser.add_argument('-r', '--resume', required=False, help='use VM snapshot from existing workdir (for cov/gdb)',
                        action='store_true', default=False)
    parser.add_argument('-p', '--processes', required=False, metavar='<n>', type=int, default=1,
                        help='number of parallel processes')
    parser.add_argument('-v', '--verbose', required=False, action='store_true', default=False,
                        help='enable verbose output')
    parser.add_argument('-q', '--quiet', help='only print warnings and errors to console',
                        required=False, action='store_true', default=False)
    parser.add_argument('-l', '--log', help='enable logging to $workdir/debug.log',
                        action='store_true', default=False)
    parser.add_argument('--debug', help='enable extra debug checks and max logging verbosity',
                        action='store_true', default=False)

# kAFL/Fuzzer-specific options
def add_args_fuzzer(parser):
    parser.add_argument('--seed-dir', required=False, metavar='<dir>', action=FullPath,
                        type=parse_is_dir, help='path to the seed directory.')
    parser.add_argument('--dict', required=False, metavar='<file>', type=parse_is_file, action=FullPath,
                        help='import dictionary file for use in havoc stage.', default=None)
    parser.add_argument('--funky', required=False, help='perform extra validation and store funky inputs.',
                        action='store_true', default=False)

    parser.add_argument('-D', '--afl-dumb-mode', required=False, help='skip deterministic stage (dumb mode)',
                        action='store_true', default=False)
    parser.add_argument('--afl-no-effector', required=False, help=hidden('disable effector maps during deterministic stage'),
                        action='store_true', default=False)
    parser.add_argument('--afl-skip-zero', required=False, help=hidden('skip zero bytes during deterministic stage'),
                        action='store_true', default=False)
    parser.add_argument('--afl-skip-range', required=False, type=parse_ignore_range, metavar="<start-end>",
                        action='append', help=hidden('skip byte range during deterministic stage'))
    parser.add_argument('--afl-arith-max', metavar='<n>', help=hidden("max arithmetic range for afl_arith_n mutation"),
                        type=int, required=False, default=35)

    parser.add_argument('--radamsa', required=False, help='enable Radamsa as additional havoc stage',
                        action='store_true', default=False)
    parser.add_argument('--grimoire', required=False, help='enable Grimoire analysis & mutation stages',
                        action='store_true', default=False)
    parser.add_argument('--redqueen', required=False, help='enable Redqueen trace & insertion stages',
                        action='store_true', default=False)
    parser.add_argument('--redqueen-hashes', required=False, help=hidden('enable Redqueen checksum fixer (broken)'),
                        action='store_true', default=False)
    parser.add_argument('--redqueen-hammer', required=False, help=hidden('enable Redqueen jump table hammering'),
                        action='store_true', default=False)
    parser.add_argument('--redqueen-simple', required=False, help=hidden('do not ignore simple matches in Redqueen'),
                        action='store_true', default=False)
    parser.add_argument('--cpu-offset', metavar='<n>', help="start CPU pinning at offset <n>",
                        type=int, default=0, required=False)
    parser.add_argument('--abort-time', metavar='<n>', help="exit after <n> hours",
                        type=float, required=False, default=None)
    parser.add_argument('--abort-exec', metavar='<n>', help="exit after max <n> executions",
                        type=int, required=False, default=None)
    parser.add_argument('-ts', '--t-soft', dest='timeout_soft', required=False, metavar='<n>', help="soft execution timeout (in seconds)",
                        type=float, default=1/1000)
    parser.add_argument('-tc', '--t-check', dest='timeout_check', required=False, help="validate timeouts against hard limit (slower)",
                        action='store_true', default=False)
    parser.add_argument('--kickstart', metavar='<n>', help="kickstart fuzzing with <n> byte random strings (default 256, 0 to disable)",
                        type=int, required=False, default=256)
    parser.add_argument('--radamsa-path', metavar='<file>', help=hidden('path to radamsa executable'),
                        type=parse_is_file, action=FullPath, required=False, default=None)


# Qemu/Worker-specific launch options
def add_args_qemu(parser):

    config_default_base   = '-enable-kvm -machine kAFL64-v1 -cpu kAFL64-Hypervisor-v1,+vmx -no-reboot -net none -display none'
    config_default_serial = '-device isa-serial,chardev=kafl_serial'
    config_default_append = 'nokaslr oops=panic nopti mitigations=off console=ttyS0'

    # BIOS/Image/Kernel load modes are partly exclusive, but we need at least one of them
    parser.add_argument('--image', dest='qemu_image', metavar='<qcow2>', required=False, action=FullPath, 
                        type=parse_is_file, help='path to Qemu disk image.')
    parser.add_argument('--snapshot', dest='qemu_snapshot', metavar='<dir>', required=False, action=FullPath,
                        type=parse_is_dir, help='path to VM pre-snapshot directory.')
    parser.add_argument('--bios', dest='qemu_bios', metavar='<file>', required=False, action=FullPath, type=parse_is_file,
                        help='path to the BIOS image.')
    parser.add_argument('--kernel', dest='qemu_kernel', metavar='<file>', required=False, action=FullPath,
                        type=parse_is_file, help='path to the Kernel image.')
    parser.add_argument('--initrd', dest='qemu_initrd', metavar='<file>', required=False, action=FullPath, type=parse_is_file,
                        help='path to the initrd/initramfs file.')
    parser.add_argument('--append', dest='qemu_append', metavar='<str>', help='Qemu -append option',
                        type=str, required=False, default=config_default_append)
    parser.add_argument('-m', '--memory', dest='qemu_memory', metavar='<n>', help='size of VM RAM in MB (default: 256).',
                        default=256, type=int)

    parser.add_argument('--qemu-base', metavar='<str>', help='base Qemu config (check defaults!)',
                        type=str, required=False, default=config_default_base)
    parser.add_argument('--qemu-serial', metavar='<str>', help='Qemu serial emulation (redirected to file, see defaults)',
                        type=str, required=False, default=config_default_serial)
    parser.add_argument('--qemu-extra', metavar='<str>', help='extra Qemu config (check defaults!)',
                        type=str, required=False, default=None)
    parser.add_argument('--qemu-path', metavar='<file>', help=hidden('path to Qemu-Nyx executable'),
                        type=parse_is_file, required=True, default=None)

    parser.add_argument('-ip0', required=False, default=None, metavar='<n-m>', type=parse_range_ip_filter,
                        help='set IP trace filter range 0 (should be page-aligned)')
    parser.add_argument('-ip1', required=False, default=None, metavar='<n-m>', type=parse_range_ip_filter,
                        help='Set IP trace filter range 1 (should be page-aligned)')
    parser.add_argument('-ip2', required=False, default=None, metavar='<n-m>', type=parse_range_ip_filter,
                        help=hidden('Set IP trace filter range 2 (should be page-aligned)'))
    parser.add_argument('-ip3', required=False, default=None, metavar='<n-m>', type=parse_range_ip_filter,
                        help=hidden('Set IP trace filter range 3 (should be page-aligned)'))

    parser.add_argument('--sharedir', metavar='<dir>', required=False, action=FullPath,
                        type=parse_is_dir, help='path to the page buffer share directory.')
    parser.add_argument('-R', '--reload', metavar='<n>', help='snapshot-reload every N execs (default: 1)',
                        type=int, required=False, default=1)
    parser.add_argument('--gdbserver', required=False, help=hidden('enable Qemu gdbserver (use via kafl_debug.py!'),
                        action='store_true', default=False)
    parser.add_argument('--log-hprintf', required=False, help="redirect hprintf logging to workdir/hprintf_NN.log",
                        action='store_true', default=False)
    parser.add_argument('--log-crashes', required=False, help="store hprintf logs only for crashes/timeouts",
                        action='store_true', default=False)
    parser.add_argument('-t', '--t-hard', dest='timeout_hard', required=False, metavar='<n>', help="hard execution timeout (seconds)",
                        type=float, default=4)
    parser.add_argument('--payload-size', metavar='<n>', help=hidden("maximum payload size in bytes (minus headers)"),
                        type=int, required=False, default=131072)
    parser.add_argument('--bitmap-size', metavar='<n>', help="size of feedback bitmap (must be power of 2)",
                        type=int, required=False, default=65536)
    parser.add_argument('--trace', required=False, help='store binary PT traces of new inputs (fast).',
                        action='store_true', default=False)
    parser.add_argument("--trace-cb", required=False, help='store decoded PT traces of new inputs (slow).',
                        action='store_true', default=False)


# kafl_debug launch options
def add_args_debug(parser):

    debug_modes = ["benchmark", "gdb", "trace", "single", "trace-qemu", "noise", "printk", "redqueen",
                   "redqueen-qemu", "verify"]
    
    debug_modes_help = '<benchmark>\tperform performance benchmark\n' \
                       '<gdb>\t\trun payload with Qemu gdbserver (must compile without redqueen!)\n' \
                       '<trace>\t\tperform trace run\n' \
                       '<trace-qemu>\tperform trace run and print QEMU stdout\n' \
                       '<noise>\t\tperform run and messure nondeterminism\n' \
                       '<printk>\t\tredirect printk calls to kAFL\n' \
                       '<redqueen>\trun redqueen debugger\n' \
                       '<redqueen-qemu>\trun redqueen debugger and print QEMU stdout\n' \
                       '<verify>\t\trun verifcation steps\n'
    
    parser.add_argument('--input', metavar='<file/dir>', action=FullPath, type=str,
                        help='path to input file or workdir.')
    parser.add_argument('-n', '--iterations', metavar='<n>', help='execute <n> times (for some actions)',
                        default=5, type=int)
    parser.add_argument('--action', required=False, metavar='<cmd>', choices=debug_modes,
                        help=debug_modes_help)
    parser.add_argument('--ptdump-path', metavar='<file>', help=hidden('path to ptdump executable'),
                        type=parse_is_file, required=True, default=None)


class ConfigArgsParser():

    def _base_parser(self):
        short_usage = '%(prog)s --work-dir <dir> [fuzzer options] [qemu options]'
        return argparse.ArgumentParser(usage=short_usage, add_help=False, fromfile_prefix_chars='@')

    def _parse_with_config(self, parser):

        config = confuse.Configuration('kafl', modname='kafl_fuzzer')

        # check default config search paths
        config.read(defaults=True, user=True)

        # local / workdir config
        workdir_config = os.path.join(os.getcwd(), 'kafl.yaml')
        if os.path.exists(workdir_config):
            config.set_file(workdir_config)

        # ENV based config
        if 'KAFL_CONFIG_FILE' in os.environ:
            config.set_file(os.environ['KAFL_CONFIG_FILE'])

        # merge all configs into a flat dictionary, delimiter = ':'
        config_values = FlatDict(config.flatten())
        if 'KAFL_CONFIG_DEBUG' in os.environ:
            print("Options picked up from config: %s" % str(config_values))

        # adopt defaults into parser, fixup 'required' and file/path fields
        for action in parser._actions:
            #print("action: %s" % repr(action))
            if action.dest in config_values:
                if action.type == parse_is_file:
                    action.default = config[action.dest].as_str_expanded()
                elif isinstance(action, argparse._AppendAction):
                    assert("append are not supported in in yaml config")
                    #action.default = [config[action.dest].as_str()]
                else:
                    action.default = config[action.dest].get()
                action.required = False
                config_values.pop(action.dest)
        
        # remove options not defined in argparse (set_defaults() imports everything)
        for option in config_values:
            if 'KAFL_CONFIG_DEBUG' in os.environ:
                logger.warn("Dropping unrecognized option '%s'." % option)
            config_values.pop(option)

        # allow unrecognized options?
        #parser.set_defaults(**config_values)

        args = parser.parse_args()

        if 'KAFL_CONFIG_DEBUG' in os.environ:
            print("Final parsed args: %s" % repr(args))
        return args

    def parse_fuzz_options(self):

        parser = self._base_parser()

        general = parser.add_argument_group('General options')
        add_args_general(general)

        fuzzer = parser.add_argument_group('Fuzzer options')
        add_args_fuzzer(fuzzer)

        qemu = parser.add_argument_group('Qemu/Nyx options')
        add_args_qemu(qemu)

        return self._parse_with_config(parser)

    def parse_debug_options(self):

        parser = self._base_parser()

        general = parser.add_argument_group('General options')
        add_args_general(general)

        debugger = parser.add_argument_group('Debug options')
        add_args_debug(debugger)

        qemu = parser.add_argument_group('Qemu options')
        add_args_qemu(qemu)

        return self._parse_with_config(parser)
